internals/overlord/cmdstate/manager.go

```diff
@@ -64,42 +66,55 @@ func (m *CommandManager) Connect(r *http.Request, w http.ResponseWriter, task *s
 		// So waitExecution wakes up if it's stuck in Wait().
 		m.executionsCond.L.Lock()
 		close(stopWait)
+		delete(m.errors, task.ID())
 		m.executionsCond.Broadcast()
 		m.executionsCond.L.Unlock()
 	}()
 
 	executionCh := make(chan *execution)
+	errCh := make(chan error)
 	go func() {
-		e := m.waitExecution(task.ID(), stopWait)
+		e, err := m.waitExecution(task.ID(), stopWait)
 		if e != nil {
 			executionCh <- e
 		}
+		if err != nil {
+			errCh <- err
+		}
 	}()
 
 	// Wait till the execution object is ready or the request is cancelled.
 	select {
 	case e := <-executionCh:
 		return e.connect(r, w, websocketID)
+	case err := <-errCh:
```

Comment:
I think you're right to add a new case here, however, I played with this yesterday, and I don't think we need any additional machinery. If we wait for the task/change to be done, that should do it, for example `case <-change.Ready()`. You can get the change using `task.Change()` (though you probably need to lock/unlock state to get the change. Then there's `change.Err()` to get the error. I've played with it locally and arrived at this:

```
	case <-change.Ready():
		// Change unexpected marked ready, probably due to client not sending
		// websocket requests in time.
		st.Lock()
		err := change.Err()
		st.Unlock()
		return fmt.Errorf("change unexpectedly marked ready: %v", err)
```

What do you think?